akka23-common/src/main/scala/com/typesafe/conductr/akka/ConnectionHandler.scala

```diff
@@ -71,38 +76,55 @@ class ConnectionHandler extends AbstractConnectionHandler {
+    payload: Option[HttpPayload])(handler: (Int, Map[String, Option[String]]) => Option[T])(implicit cc: CC): Future[Option[T]] = {
 
     payload.fold[Future[Option[T]]](Future.successful(None)) { p =>
-      val connection = cc.httpExt.outgoingConnection(p.getUrl.getHost, p.getUrl.getPort)
-      val url = p.getUrl
-      val urlStr = url.toString
-
-      val requestMethod = p.getRequestMethod match {
-        case "GET"     => Get(urlStr)
-        case "POST"    => Post(urlStr)
-        case "PUT"     => Put(urlStr)
-        case "PATCH"   => Patch(urlStr)
-        case "DELETE"  => Delete(urlStr)
-        case "OPTIONS" => Options(urlStr)
-        case "HEAD"    => Head(urlStr)
-      }
-
-      val request = requestMethod
-        .addHeader(`User-Agent`(UserAgent))
-        .addHeader(Host(url.getHost))
+      val connection = createConnection(p, cc)
+      val request = createRequest(p)
 
       val requestSource = Source.single(request)
         .via(connection)
-        .map { response =>
-          thunk(
-            response.status.intValue(),
-            response.headers.foldLeft(Map.empty[String, Option[String]]) {
-              case (m, header) => m.updated(header.name(), Some(header.value()))
-            })
-        }
+        .map(response => handler(response.status.intValue(), transformHeaders(response.headers)))
       requestSource.runWith(Sink.head)(cc.actorMaterializer)
     }
+  }
+
+  // FIXME: Need to be abstracted in AbstractConnectionHandler. Problem is that this method is not needed in the conductr-bundle-lib modules.
+  def withConnectedRequest[T](payload: HttpPayload)(handler: (Int, ResponseEntity) => Future[T])(implicit cc: CC): Future[T] = {
+    val connection = createConnection(payload, cc)
+    val request = createRequest(payload)
 
+    Source
+      .single(request)
+      .via(connection)
+      .mapAsync(1)(response => handler(response.status.intValue(), response.entity))
+      .runWith(Sink.head)(cc.actorMaterializer)
   }
+
+  private def createConnection(payload: HttpPayload, cc: CC) =
+    cc.httpExt.outgoingConnection(payload.getUrl.getHost, payload.getUrl.getPort)
+
+  private def createRequest(payload: HttpPayload): HttpRequest = {
+    val url = payload.getUrl
+    val urlStr = url.toString
+
+    val requestMethod = payload.getRequestMethod match {
+      case "GET"     => Get(urlStr)
+      case "POST"    => Post(urlStr)
+      case "PUT"     => Put(urlStr)
+      case "PATCH"   => Patch(urlStr)
+      case "DELETE"  => Delete(urlStr)
+      case "OPTIONS" => Options(urlStr)
+      case "HEAD"    => Head(urlStr)
+    }
+
+    requestMethod
+      .addHeader(`User-Agent`(UserAgent))
+      .addHeader(Host(url.getHost))
+  }
+
+  private def transformHeaders(headers: Seq[HttpHeader]): Map[String, Option[String]] =
+    headers.foldLeft(Map.empty[String, Option[String]]) {
+      case (m, header) => m.updated(header.name(), Some(header.value()))
+    }
```

Comment:
We have now two `withConnectedRequest()` methods with a different signature. Therefore the code which is shared across these two methods is encapsulated.